libp2p/connection.nim

```diff
@@ -0,0 +1,98 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos
+import peerinfo, multiaddress, stream/lpstream, peerinfo, varint, vbuffer
+
+const DefaultReadSize: uint = 64*1024
+
+type
+  Connection* = ref object of LPStream
+    peerInfo*: PeerInfo
+    stream*: LPStream
+
+proc newConnection*(stream: LPStream): Connection =
+  ## create a new Connection for the specified async reader/writer
+  new result
+  result.stream = stream
+
+method read*(s: Connection, n = -1): Future[seq[byte]] {.gcsafe.} =
+  result = s.stream.read(n)
+
+method readExactly*(s: Connection,
+                    pbytes: pointer,
+                    nbytes: int): Future[void] {.gcsafe.} =
+  result = s.stream.readExactly(pbytes, nbytes)
+
+method readLine*(s: Connection,
+                 limit = 0,
+                 sep = "\r\n"): Future[string] {.gcsafe.} =
+  result = s.stream.readLine(limit, sep)
+
+method readOnce*(s: Connection,
+                 pbytes: pointer,
+                 nbytes: int): Future[int] {.gcsafe.} =
+  result = s.stream.readOnce(pbytes, nbytes)
+
+method readUntil*(s: Connection,
+                  pbytes: pointer,
+                  nbytes: int,
+                  sep: seq[byte]): Future[int] {.gcsafe.} =
+  result = s.stream.readUntil(pbytes, nbytes, sep)
+
+method write*(s: Connection, pbytes: pointer, nbytes: int): Future[void] {.gcsafe.} =
+  result = s.stream.write(pbytes, nbytes)
+
+method write*(s: Connection, msg: string, msglen = -1): Future[void] {.gcsafe.} =
+  result = s.stream.write(msg, msglen)
+
+method write*(s: Connection, msg: seq[byte], msglen = -1): Future[void] {.gcsafe.} =
+  result = s.stream.write(msg, msglen)
+
+method close*(s: Connection) {.async, gcsafe.} =
+  await s.stream.close()
+  s.closed = true
+
+proc readLp*(s: Connection): Future[seq[byte]] {.async, gcsafe.} =
+  ## read lenght prefixed msg
```

Comment:
This is a facade class that forwards calls to some underlying implementation. This is done so that I can compose concrete implementations. Admittedly, I could have put my helper Read/WriteLp methods under LPSteam, but they feel a bit out of place, since LPStream shouldn't know anything about varints and length prefixed data. Not opposed to revisit it once there is a more concrete implementation.